### PR TITLE
Test MediaType negotiation

### PR DESCRIPTION
**Pull request by shiliyu**
1. Fix `produces` and `consumes` match in `PlainHttpRoutingMatchResult.java`.
2. Add test for `405` `406` `415` scenario.
3. Fix IDE warnings and ktlint failure.

### CODE REVIEW

File: core/src/main/java/io/forestframework/core/http/routing/PlainHttpRoutingMatchResult.java [hunk 1]
```diff
@@ -42,8 +42,8 @@ public void select(HttpServerRequest request,
 
     public void addResult(HttpServerRequest request, List<Routing> routings, Map<String, String> pathVariables) {
         routings.stream()
-                .collect(Collectors.groupingBy(Routing::getType))
-                .forEach((key, value) -> putIntoResult(request, key, value, pathVariables));
+            .collect(Collectors.groupingBy(Routing::getType))
```
Comment (blindpirate) [inline core/src/main/java/io/forestframework/core/http/routing/PlainHttpRoutingMatchResult.java:45]: Can you also add rules to `editorconfig` so we don't have these space changes in the future?

File: core/src/test/java/io/forestframework/core/http/routing/AbstractHttpIntegrationTest.java [hunk 2]
```diff
@@ -21,7 +23,14 @@
...
     public HttpResponse sendHttpRequest(String method, String path) throws IOException {
         String uri = "http://localhost:" + port + path;
-        HttpUriRequest request = RequestBuilder.create(method).setUri(uri).build();
+        HttpUriRequest request = RequestBuilder
+                .create(method)
+                .setUri(uri)
+                .setHeader("Accept", String.valueOf(ContentType.APPLICATION_JSON))
+                .setHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+//                .setHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36")
+//                .setHeader(HttpHeaders.CONTENT_TYPE, "text/html")
```
Comment (blindpirate) [inline core/src/test/java/io/forestframework/core/http/routing/AbstractHttpIntegrationTest.java]: Either remove the comment or uncomment, why do you leave the comments here?

File: core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package io.forestframework.core.http.routing;
```
Comment (blindpirate) [inline core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java]: Unused class.

File: core/src/test/java/io/forestframework/core/http/routing/Message.java [hunk 4]
```diff
@@ -12,6 +12,8 @@
     CUSTOM_ERROR_HANDLER,
     CUSTOM_404_ERROR_HANDLER,
     CUSTOM_4XX_ERROR_HANDLER,
+    CUSTOM_405_ERROR_HANDLER,
+    CUSTOM_406_ERROR_HANDLER,
```
Comment (blindpirate) [inline core/src/test/java/io/forestframework/core/http/routing/Message.java]: They're not used.

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    val httpMethod: HttpMethod = HttpMethod.GET
+)
+
+// Put your test cases here
+val producesAcceptCases = listOf(
+    ProducesAcceptCase("get 200 when matching single", produces = listOf("application/json"), accept = "application/json", expectedStatusCode = 200),
+    ProducesAcceptCase("get 406 when matching single", produces = listOf("application/json"), accept = "text/html", expectedStatusCode = 406),
+
+    ProducesAcceptCase("get 200 when matching multiple", produces = listOf("application/json", "text/plain"), accept = "text/html, application/json;q=0.9, text/plain;q=0.8", expectedStatusCode = 200),
+    ProducesAcceptCase("get 406 when matching multiple", produces = listOf("application/json", "text/plain"), accept = "application/xhtml+xml, text/html;q=0.3", expectedStatusCode = 406),
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: `get 406 when matching multiple` or `not matching`?

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    ProducesAcceptCase("get 200 when matching wildcard match", produces = listOf("text/*"), accept = "text/html, application/json; level=1;q=0.9, text/plain;q=0.8", expectedStatusCode = 200),
+    ProducesAcceptCase("get 406 when matching wildcard match", produces = listOf("application/json"), accept = "text/*", expectedStatusCode = 406),
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: `matching`? `not matching`?

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    ProducesAcceptCase("get 200 when matching double wildcard case 1", produces = listOf("*/*"), accept = "*/*;level=1;q=0.9", expectedStatusCode = 200),
+    ProducesAcceptCase("get 200 when matching double wildcard case 2", produces = listOf("text/*"), accept = "*/*;q=0.7", expectedStatusCode = 200),
+    ProducesAcceptCase("get 200 when matching double wildcard case 3", produces = listOf("*/*"), accept = "text/*", expectedStatusCode = 200),
+    ProducesAcceptCase("get 200 when matching double wildcard case 4", produces = listOf("*/*"), accept = "application/xhtml+xml;q=0.7, image/jxr;level=1", expectedStatusCode = 200),
+
+    ProducesAcceptCase("get 200 when matching with parameter", produces = listOf("application/json;charset=UTF-8"), accept = "application/json;q=0.3", expectedStatusCode = 200),
+    ProducesAcceptCase("get 406 when matching with parameter", produces = listOf("application/json;charset=UTF-8"), accept = "application/json;level=1;charset=UTF-16", expectedStatusCode = 406)
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: `matching`? `not matching`?

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+val consumesContentTypeCases = listOf(
+    ConsumesContentTypeCase("get 200 when matching single", consumes = listOf("application/json"), contentType = "application/json", expectedStatusCode = 200),
+    ConsumesContentTypeCase("get 415 when matching single", consumes = listOf("application/json"), contentType = "text/html", expectedStatusCode = 415),
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: `matching`? `not matching`?

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+            val httpClient = vertx.createHttpClient()
+            httpClient.get(port, "/test", mapOf("Content-Type" to case.contentType)).assertStatusCode(case.expectedStatusCode)
+        }
+    }
+
+    @TestFactory
+    fun testMethodNotAllowed() = methodNotAllowedCases.map(this::testTwo)
+
+    private fun testTwo(case: MethodNotAllowedCase) = DynamicTest.dynamicTest(case.description) {
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]:
```suggestion
    private fun testOne(case: MethodNotAllowedCase) = DynamicTest.dynamicTest(case.description) {
```

`testOne` means "test one case", not "test1"

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    ConsumesContentTypeCase("get 200 when matching double wildcard case 1", consumes = listOf("*/*"), contentType = "application/json", expectedStatusCode = 200),
+    ConsumesContentTypeCase("get 200 when matching double wildcard case 2", consumes = listOf("*/*"), contentType = "text/*", expectedStatusCode = 200),
+
+    ConsumesContentTypeCase("get 200 when matching with parameter", consumes = listOf("application/json;charset=UTF-8"), contentType = "application/json;q=0.9", expectedStatusCode = 200),
+    ConsumesContentTypeCase("get 415 when matching with parameter", consumes = listOf("application/json;charset=UTF-8"), contentType = "application/json;charset=UTF-16", expectedStatusCode = 415)
+)
+
+val methodNotAllowedCases = listOf(
+    MethodNotAllowedCase("get 405 if method not allowed", httpMethod = HttpMethod.POST, expectedStatusCode = 405)
```
Comment (blindpirate) [inline core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: Several more test cases for:

- httpMethod = ALL
- httpMethod = GET and POST, request http method = GET
- httpMethod = GET and POST, request http method = DELETE or HEAD

**Review by blindpirate — CHANGES_REQUESTED** [review]
Much better now.

File: core/src/main/java/io/forestframework/core/http/routing/PlainHttpRoutingMatchResult.java (hunk 1 above)
Comment (shiliyu) [reply core/src/main/java/io/forestframework/core/http/routing/PlainHttpRoutingMatchResult.java:45]: Sure!

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt (hunk 5 above)
Comment (shiliyu) [reply core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: fix it right away!

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt (hunk 9 above)
Comment (shiliyu) [reply core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: 'kay, got it.

File: core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt (hunk 10 above)
Comment (shiliyu) [reply core/src/test/kotlin/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.kt]: Sure!

File: core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java (hunk 3 above)
Comment (shiliyu) [reply core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java]: Do you think it's better to remove `MediaTypeNegotiatingIntegrationTest.java`? it's never used.

File: core/src/test/java/io/forestframework/core/http/routing/AbstractHttpIntegrationTest.java (hunk 2 above)
Comment (shiliyu) [reply core/src/test/java/io/forestframework/core/http/routing/AbstractHttpIntegrationTest.java]: i just forgot to delete 'em. it's not a good practice.

File: core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java (hunk 3 above)
Comment (blindpirate) [reply core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java]: yes, please.

Comment (shiliyu) [reply core/src/test/java/io/forestframework/core/http/routing/MediaTypeNegotiatingIntegrationTest.java]: will do.

File: test-fixtures/src/main/kotlin/io/forestframework/testfixtures/AbstractForestIntegrationTest.kt [hunk 11]
```diff
@@ -87,6 +88,13 @@ suspend fun HttpClient.get(port: Int, uri: String, headers: Map<String, String>
     }
 }
 
+suspend fun HttpClient.send(method: HttpMethod, port: Int, uri: String) = awaitResult<HttpClientResponse> { handler ->
+    send(method, port, "localhost", uri) {
```
Comment (shiliyu) [inline test-fixtures/src/main/kotlin/io/forestframework/testfixtures/AbstractForestIntegrationTest.kt:92]: Could you plz help implement this function? it's beyond me. @blindpirate

File: test-fixtures/src/main/kotlin/io/forestframework/testfixtures/AbstractForestIntegrationTest.kt [hunk 12]
```diff
@@ -88,10 +88,10 @@ suspend fun HttpClient.get(port: Int, uri: String, headers: Map<String, String>
     }
 }
 
-suspend fun HttpClient.send(httpMethod: HttpMethod, port: Int, uri: String, headers: Map<String, String> = emptyMap()) = awaitResult<HttpClientResponse> { handler ->
+suspend fun HttpClient.delete(port: Int, uri: String, headers: Map<String, String> = emptyMap()) = awaitResult<HttpClientResponse> { handler ->
```
Comment (blindpirate) [inline test-fixtures/src/main/kotlin/io/forestframework/testfixtures/AbstractForestIntegrationTest.kt]:
```
suspend fun HttpClient.delete(port: Int, uri: String, headers: Map<String, String> = emptyMap()) = send(DELETE, port, uri, headers)

suspend fun HttpClient.get(port: Int, uri: String, headers: Map<String, String> = emptyMap()) = send(GET, port, uri, headers)
```

Comment (shiliyu) [reply test-fixtures/src/main/kotlin/io/forestframework/testfixtures/AbstractForestIntegrationTest.kt]: oh... i didn't realize it until right now

**Review by blindpirate — APPROVED** [review]
Nice work!

**Pull request merged**